src/middlewared/middlewared/plugins/auth.py

```diff
@@ -291,13 +295,25 @@ def get_token(self, token_id):
             return None
 
     @no_auth_required
-    @accepts(Str('username'), Str('password'))
+    @accepts(Str('username'), Str('password'), Str('otp_token', null=True, default=None))
     @pass_app
-    async def login(self, app, username, password):
-        """Authenticate session using username and password.
+    async def login(self, app, username, password, otp_token=None):
+        """
+        Authenticate session using username and password.
         Currently only root user is allowed.
+        `otp_token` must be specified if two factor authentication is enabled.
         """
         valid = await self.check_user(username, password)
+        twofactor_auth = await self.middleware.call('auth.twofactor.config')
+
+        if twofactor_auth['enabled']:
+            # We should validate that BOTH `check_user` AND `auth.twofactor.verify` are true and only then
```

Comment:
That's not the real reason, isn't it? The real reason is to prevent guessing passwords with timing attack.

Reply:
Yes indeed. That's what the two factor auth enabled check represents ;) I'll update the comment :)